online-shop/src/app/components/presentational/products-form-view/products-form-view.component.ts

```diff
@@ -0,0 +1,114 @@
+import {
+  Component,
+  EventEmitter,
+  Input,
+  OnChanges,
+  OnInit,
+  Optional,
+  Output,
+  SimpleChanges,
+} from '@angular/core';
+import {
+  FormControl,
+  FormGroup,
+  Validators
+} from '@angular/forms';
+import { ProductAndProductCategory } from 'src/app/types/productsAndProductsCategory';
+
+@Component({
+  selector: 'app-products-form-view',
+  templateUrl: './products-form-view.component.html',
+  styleUrls: ['./products-form-view.component.scss'],
+})
+export class ProductsFormViewComponent implements OnInit, OnChanges {
+  @Input() @Optional() product: ProductAndProductCategory | undefined;
+  @Output() @Optional() updateProductEvent = new EventEmitter<ProductAndProductCategory>();
+  @Output()@Optional() addProductEvent = new EventEmitter<ProductAndProductCategory>();
+  myForm!: FormGroup;
+
+
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if (this.product) {
+      this.myForm.patchValue({
+        name: this.product.productName,
+        category: this.product.productCategoryName,
+        image: this.product.productImageUrl,
+        price: this.product.productPrice,
+        description: this.product.productDescription,
+      });
+    }
+  }
+
+  ngOnInit(): void {
+    this.myForm = new FormGroup({
+      name: new FormControl('', [Validators.required, Validators.minLength(4)]),
+      category: new FormControl('', [
+        Validators.required,
+        Validators.minLength(4),
+      ]),
+      image: new FormControl('', [
+        Validators.required,
+        Validators.minLength(4),
+        this.validateImageUrl,
+      ]),
+      price: new FormControl('', [
+        Validators.required,
+        Validators.pattern(/^\d+(\.\d{1,2})?$/),
+      ]),
+      description: new FormControl('', [
+        Validators.required,
+        Validators.minLength(10),
+      ]),
+    });
+  }
+  handleSave() {
+    if (this.product) this.editProductHandle();
+    else this.addProductHandle();
+  }
+
+  addProductHandle(){
```

Comment:
In general, as convention, "handle" is put first, then the rest of the method: "handleAddProduct", "handleCheckout", "handleSave" etc